design-proposals/klip-32-sql-testing-tool.md

```diff
@@ -0,0 +1,284 @@
+psuedocolumns as well.
+
+#### `ASSERT NO DATA`
+
+The `ASSERT NO DATA` statement will allow users to ensure that no more data exists in a specified
+source:
+```sql
+ASSERT NO DATA source_name [OFFSET from_offset];
+```
+
+#### `ASSERT SOURCE`
+
+The `ASSERT (TABLE | STREAM)` statement asserts that the given stream or table has the specified
+columns and physical properties.
+```sql
+ASSERT (STREAM | TABLE) source_name ( { column_name data_type [[PRIMARY] KEY] } [, ...] )
+    WITH (property_name = expression [, ...] );
+```
+
+#### `ASSERT TYPE`
+
+The `ASSERT TYPE` statement ensures that a custom type has the expected type. This is especially
+useful when chaining multiple `CREATE TYPE` statements together and asserting the types later
+in the chain are correct.
+```sql
+ASSERT TYPE type_name AS type;
+```
+
+#### Considered Alternative
+
+It is possible to use a meta directive (see below) in order to drive asserts as well. This was
+considered and rejected because it would require additional parsing and more testing-specific code
+to convert to expressions and pass into the codegen, which diverges from the motivating principle 
+of minimizing testing-only code.
+
+### Meta Directives
+
+Some functionality cannot be addressed in standard SQL syntax and will instead be supported using
+meta directives. Some examples of these directives:
+
+- The `--@expected.error` and `--@expected.message` directives will indicate that the below test 
+    expects an exception to be thrown with the specified message and type.
+- The `--@topic.denylist` directive will ensure that any topics matching the denylist will not
+    be present at the end of the test execution.
+
+### Global Checks
+
+These are a set of checks that will happen without any directive or `ASSERT` statement. Some
+checks that we may consider including:
+
+- _Processing Log Check_: this check will ensure that there are no failures in the processing log.
+    If the test case intends to check that the processing log contains certain entries, this check
+    can be disabled on a test-by-test basis and assert via `ASSERT DATA` on the processing log
+    stream.
+    
+### Language Gaps
+
+There are some features that we need to implement in ksqlDB's SQL language in order to put YATT
+on par with the existing testing tools. Some of those are outlined here:
+
+- _Windowed Keys_: when we have structured key support, we can `ASSERT DATA` and just provide the
+    struct as the value. Until then, windowed keys will either be unsupported or stringified.
+- _Tombstones_: we will implement `DELETE FROM table WHERE key = value` syntax to allow tombstones
+    to be inserted into tables, both for ksqlDB and YATT.
+- _Null Values for Streams_: insert `null` values into streams will not be supported for the first
+    iteration of YATT. Eventually, if we want to support this, we can add a directive like
+    `--@null.value: topic key` to produce a `null` valued record into a specific topic/source.
```

Comment:
Is this not something we want to eventually support outside of testing purposes? Either (1) there are no valid use cases for inserting nulls into streams, in which case why are we testing it, or (2) there are valid use cases for inserting nulls into streams, and we should support it in ksqlDB.

Reply:
That's a good question, but even if we decide there are no valid use cases for insert nulls into streams it's still valid to test what happens if your source (bad) data has nulls in your streams.

Reply:
That's fair but it seems weird to add special support for this particular case of "bad data" in the testing tool, since there are other types of bad data we can't produce using SQL syntax either (garbage bytes, malformed data, etc). The KLIP already contains a proposal for BYO Kafka to allow for testing these scenarios. Why is the case of null values in streams special / different?

Reply:
the big one is parity with QTT/ksql-test-runner; they can't produce garbage bytes or malformed data either. BYOK is a pretty heavy extension if we wanted to bring this up to parity